### PR TITLE
Enforce LooseCoupling Rule in PMD for production code #7084

### LINKED ISSUE #7084: Enforce LooseCoupling Rule in PMD for production code.
> One of the principles in designing API is hiding of the internal implementation to users. In other words, we should avoid using implementation types and use the interface instead. There is a `PMD` rule for this: `LooseCoupling`. This rule can be applied to TEAMMATES production code.
> 
> ---- From Technical Report: Static Analysis (Java) in TEAMMATES in https://xpdavid.github.io/CS2103R-Report/

image
For example, we should expose `Map` instead of the implementation `HashMap` here.

Two tasks here:
1. Add `<rule ref="rulesets/java/typeresolution.xml/LooseCoupling"/>` to `teammates-pmdMain.xml`
2. Fix violations reported by PMD.

### PR DESCRIPTION
**Pull request by wkk91193**
Fixes #7084

### CODE REVIEW

**Comment by wkk91193** [comment]
work-in-progress

**Comment by wkurniawan07** [comment]
@wkk91193, please do not introduce unnecessary changes such as whitespace-only changes on this PR. Take a look at [the changes you introduced in this PR](https://github.com/TEAMMATES/teammates/pull/7122/files) and make sure they are all accounted for.

Other than that, it appears that you have included changes from other issue (e.g. in commit a8492f3), which should not be the case.

**Comment by wkk91193** [comment]
Ready for review

File: static-analysis/teammates-pmdMain.xml [hunk 1]
```diff
@@ -13,7 +13,7 @@
     </rule>
 
     <rule ref="rulesets/java/strictexception.xml/SignatureDeclareThrowsException"/>
-
+	<rule ref="rulesets/java/typeresolution.xml/LooseCoupling"/>
```
Comment (xpdavid) [inline static-analysis/teammates-pmdMain.xml]: Use space instead of tab here

File: src/main/java/teammates/logic/core/CoursesLogic.java [hunk 2]
```diff
@@ -555,7 +555,7 @@ public CourseSummaryBundle getCourseSummaryWithoutStats(String courseId) throws
      * @return HashMap with courseId as key, and CourseDetailsBundle as value.
```
Comment (xpdavid) [inline src/main/java/teammates/logic/core/CoursesLogic.java]: Remember to update the Javadoc also. Check the rest.

File: src/main/java/teammates/logic/core/CoursesLogic.java [hunk 3]
```diff
@@ -538,7 +538,7 @@ public CourseSummaryBundle getCourseSummaryWithoutStats(String courseId) throws
...
```
Comment (xpdavid) [inline src/main/java/teammates/logic/core/CoursesLogic.java]: Remember to update the Javadoc also. Check the rest.

**Comment by wkk91193** [comment]
@xpdavid changes made. let me know if there's any left  :)

**Comment by wkk91193** [comment]
@xpdavid ready for review

File: src/main/java/teammates/common/util/HttpRequestHelper.java [hunk 4]
```diff
@@ -36,9 +36,9 @@ public static String getValueFromParamMap(Map<String, String[]> paramMap, String
     }
 
     /**
-     * Returns a HashMap object containing all the parameters key-value pairs from a URLFetchRequest object.
+     * Returns a Map object containing all the parameters key-value pairs from a URLFetchRequest object.
```
Comment (xpdavid) [inline src/main/java/teammates/common/util/HttpRequestHelper.java]: `Returns a map <del>object</del> containing ...`
Map is an interface, technically, not object.

File: static-analysis/teammates-pmdMain.xml [hunk 5]
```diff
@@ -11,9 +11,9 @@
         <!-- Not worth the added safety -->
         <exclude name="GuardLogStatementJavaUtil"/>        
     </rule>
-
+
```
Comment (xpdavid) [inline static-analysis/teammates-pmdMain.xml]: You may want to remove the trailing spaces

File: src/main/java/teammates/common/util/HttpRequestHelper.java (hunk 4 above)
Comment (wkk91193) [inline src/main/java/teammates/common/util/HttpRequestHelper.java]: renamed "object" to "interface"

File: static-analysis/teammates-pmdMain.xml (hunk 5 above)
Comment (wkk91193) [inline static-analysis/teammates-pmdMain.xml]: Removed trailing spaces

**Comment by wkk91193** [comment]
@xpdavid The travis build failing prompting "Could not get resource 'https://repo1.maven.org/maven2/com/mailjet/mailjet-client/4.0.5/mailjet-client-4.0.5.pom" what could be causing this?

**Comment by wkk91193** [comment]
@xpdavid  the failing travis build job is now passed! Ready for review!

File: src/main/java/teammates/common/util/HttpRequestHelper.java [hunk 6]
```diff
@@ -36,9 +36,9 @@ public static String getValueFromParamMap(Map<String, String[]> paramMap, String
     }
 
     /**
-     * Returns a HashMap object containing all the parameters key-value pairs from a URLFetchRequest object.
+     * Returns a Map interface containing all the parameters key-value pairs from a URLFetchRequest object.
```
Comment (xpdavid) [inline src/main/java/teammates/common/util/HttpRequestHelper.java]: No need interface. Here
`Returns a Map containing all the parameters key-value .....`

File: static-analysis/teammates-pmdMain.xml [hunk 7]
```diff
@@ -6,14 +6,12 @@
...
-
     <rule ref="rulesets/java/logging-java.xml">
         <!-- Not worth the added safety -->
         <exclude name="GuardLogStatementJavaUtil"/>        
-    </rule>
-
+    </rule>    
     <rule ref="rulesets/java/strictexception.xml/SignatureDeclareThrowsException"/>
-
+    <rule ref="rulesets/java/typeresolution.xml/LooseCoupling"/>
```
Comment (xpdavid) [inline static-analysis/teammates-pmdMain.xml]: Pls discard the rest of changes in `teammates-pmdMain.xml`. Only need to keep the changes for this line.

**Comment by wkk91193** [comment]
@xpdavid  Ready for review!

File: src/main/java/teammates/common/util/HttpRequestHelper.java [hunk 8]
```diff
@@ -32,6 +36,30 @@ public static String getValueFromParamMap(Map<String, String[]> paramMap, String
     }
 
     /**
+     * Returns a Map containing all the parameters key-value pairs from a URLFetchRequest object.
+     */
+    public static Map<String, String> getParamMap(URLFetchRequest request) {
```
Comment (xpdavid) [inline src/main/java/teammates/common/util/HttpRequestHelper.java]: Why do you need the method?

**Comment by wkurniawan07** [comment]
Hi @wkk91193, any updates here?

**Comment by wkk91193** [comment]
@wkurniawan07 @xpdavid ready for review!

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java [hunk 9]
```diff
@@ -1709,7 +1709,7 @@ public String getQuestionText(String feedbackQuestionId) {
         return sortedMap;
     }
 
-    public LinkedHashMap<String, Map<FeedbackQuestionAttributes, List<FeedbackResponseAttributes>>>
+    public Map<String, Map<FeedbackQuestionAttributes, List<FeedbackResponseAttributes>>>
```
Comment (xpdavid) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: Add a Javadoc to state that the method returns a ordered Map with {@code recipientTeam} name as key sorted by {@code compareByTeamQuestionRecipientTeamGiver} comparator.

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java [hunk 10]
```diff
@@ -1742,7 +1742,7 @@ public String getQuestionText(String feedbackQuestionId) {
...
```
Comment (xpdavid) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: Same here.

**Comment by wkk91193** [comment]
@xpdavid  Ready for review!

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java [hunk 11]
```diff
@@ -1742,7 +1746,11 @@ public String getQuestionText(String feedbackQuestionId) {
         return sortedMap;
     }
 
-    public LinkedHashMap<String, Map<FeedbackQuestionAttributes, List<FeedbackResponseAttributes>>>
+    /**
+     * Returns an ordered Map with {@code giverTeam} name as key
+     * sorted by {@code compareByTeamQuestionGiverTeamRecipient} comparator.
```
Comment (xpdavid) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: Take a look here
``` java
         if (sortByTeam) {
             Collections.sort(responses, compareByTeamRecipientGiverQuestion);
         } else {
             Collections.sort(responses, compareByRecipientGiverQuestion);
         }
```

The comparator is `compareByTeamRecipientGiverQuestion ` or `compareByRecipientGiverQuestion `
Sorry I didn't make it clear in my previous review

Comment (wkk91193) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: @xpdavid Pardon! the point you're trying to explain is not very clear to me

Comment (xpdavid) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: In this method, it is not sorted by "compareByTeamQuestionGiverTeamRecipient", rather it is `compareByTeamRecipientGiverQuestion` or `compareByRecipientGiverQuestion `

Comment (xpdavid) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: @wkk91193 Reminder. You are almost there. :)

**Comment by wkk91193** [comment]
@xpdavid  ready for review!

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java (hunk 11 above)
Comment (xpdavid) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: @wkk91193 Correct me If I am wrong. I think I have made a mistake. :(
This method is still using `compareByTeamQuestionGiverTeamRecipient` right?

Comment (wkk91193) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: @xpdavid we're all are human :) by the way I reverted to a previous commit but Travis CI had not reported the status yet! Is there anything wrong?

Comment (xpdavid) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: Try `git push --force` to trigger another build. Something wrong with Travis CI.

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java [hunk 12]
```diff
@@ -1709,7 +1709,11 @@ public String getQuestionText(String feedbackQuestionId) {
         return sortedMap;
     }
 
-    public LinkedHashMap<String, Map<FeedbackQuestionAttributes, List<FeedbackResponseAttributes>>>
+    /**
+     * Returns an ordered Map with {@code recipientTeam} name as key
+     * sorted by {@code compareByTeamQuestionRecipientTeamGiver} comparator.
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: `compareByTeamQuestionRecipientTeamGiver` is a private variable. It doesn't make sense to explain a public method using a private variable. Perhaps explain the sorting order in plain English? @xpdavid what do you think?

Comment (xpdavid) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: Agree. However, the sorting logic is a little bit complicated and it is not easy to describe it briefly. I was expecting the developers to look at the comparator directly but the comparator is private. Maybe we could give a high-level description like the comment for the sorter `// Sorts by recipientTeam > question > recipientName > giverTeam > giverName`.

Comment (damithc) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: Yes, that could work 👍

Comment (xpdavid) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: @wkk91193 Pls do the changes accordingly :)

**Comment by wkk91193** [comment]
@xpdavid ready for review!

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java (hunk 11 above)
Comment (xpdavid) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: This is also a private comparator. Change to `... sorted by giverTeam > question > giverName > recipientTeam > recipientName` @wkk91193

**Comment by xpdavid** [comment]
@wkk91193 Pls solve the conflict so that we can merge :)

**Comment by wkk91193** [comment]
@xpdavid, Unlike any other merge conflict, I could not find the exact location for conflict marked with conflict markers <<<<<<<, =======, >>>>>>> . Any idea on this?

**Comment by xpdavid** [comment]
As long as there is no error when you merge from master, you can just mark the conflict as resolved.

**Comment by xpdavid** [comment]
@wkk91193 Can you resolve the conflicts?

**Comment by wkk91193** [comment]
@xpdavid  ready!
